man/man8/zpoolconcepts.8

```diff
@@ -93,6 +93,56 @@ compromised.
 The minimum number of devices in a raidz group is one more than the number of
 parity disks.
 The recommended number is between 3 and 9 to help increase performance.
+.It Sy draid , draid1 , draid2 , draid3
+A distributed version of raidz with integrated hot spares that allows for
+faster resilvering.
+A draid vdev is constructed from multiple internal raidz groups, each with D
+data devices and P parity devices, which are distributed over the children.
+This allows all of the children to participate when resilvering.
+As a general rule the more devices in a draid the less time required to
+resilver to a distributed spare.
+Furthermore, a narrower stripe width (D+P) will improve resilver performance
+at the expense of total usable capacity.
+.Pp
+Like raidz a draid can have single-, double-, or triple-parity.  The
+.Sy draid1 ,
+.Sy draid2 ,
+and
+.Sy draid3
+types can be used to specified the required parity level.
+The
+.Sy draid
+vdev type is an alias for
+.Sy draid1 .
+.Pp
+A draid with N disks of size X, D data disks, P parity level, and H distributed
+hot spares can hold approximately (N-H)*(D/(D+P))*X bytes and can withstand P
+device(s) failing before data integrity is compromised.
+The minimum number of devices in a draid is P+2.
+.It Sy draid[<parity>][:<data>d][:<children>c][:<spares>s][:<iterations>i]
+When a non-default draid configuration is required it can be specified by
+appending one or more of the following optional arguments to the
+.Sy draid
+keyword.
+.Pp
+.Em children
+- The expected number of children.
+If the provided number of children is different an error is returned.
+.Pp
+.Em data
+- The number of data devices per redundancy group.
+.Pp
+.Em iterations
+- Iterations to perform when developing the permutation map.
```

Comment:
Why would I specify more or less?  What's the default?

Reply:
To keep `zpool create` fast we provide an `/etc/zfs/draid.d/default.gz` file which contains pre-computed balanced permutation mappings for the expected common configurations (but not all configurations).  These layouts are preferred if there's one available for the specified configuration.  However, when one isn't it needs to be generated on the fly.

This value controls how many candidate permutation mapping will be considered and evaluated before selecting one to be used for the pool.  Larger values can result in a more fair layout at the expense of long `zpool create` execution times for marginal gain.  This is particularly the case when the pool has a large number of children.  The default number of iterations is 1.

This feels like a lot to explain here but I'll see if I can expand on it a bit.  This isn't an option I expect most administrators to need unless they're creating pools with >32 children.  We provide mapping for all layouts smaller than this.

Reply:
I think that almost your entire explanation would be needed to understand when to specify "iterations", and what potential values to try.  Alternatively, we could have ZFS do something reasonable in all cases, remove the setting, and perhaps have it as a module parameter?  How long are we talking about to calculate a good layout?  Would it be good enough to try for 10 seconds?

Reply:
Yes, I'm not necessarily against making this a less visible option.    Originally, I even omitted it from the man page.  But in the end I added it since I think some sites (and vendors) interested in draid will want very large pools and the ability to create optimized maps for their hardware.  This can be done with the `zpool create -n` (dryrun) option which saves the mapping and prints its score for evaluation.  This is the primary use for the iterations option since you can use it to generate optimized configurations for future use.  Perhaps I can add this explanation to the explanation of the `-n` option.

From what I've seen while testing even a single iteration produces something which is reasonable  When there are a small number of child  vdevs <10 it takes perhaps a second to calculate.  For larger pools ~100 vdevs it can take several minutes.  So by default with one iteration we always do something reasonable, but it may take a few minutes to complete.

Reply:
Switching to a "fixed" algorithm with carefully selected seeds removed the need for user space utilities to be aware of the permutation maps entirely.  I've removed this option entirely since it's no longer needed.